taichi/codegen/codegen_opengl.cpp

```diff
@@ -521,11 +532,55 @@ class KernelGen : public IRVisitor {
       emit("int _tid = int(gl_GlobalInvocationID.x);");
       emit("if (_tid >= {}) return;", num_threads_);
       emit("int _itv = {} + _tid * {};", begin_value, 1 /* stmt->step? */);
-    } else {
-      TI_ERROR("[glsl] non-const range_for currently unsupported under OpenGL");
+      stmt->body->accept(this);
+    } else {{
+
+      ScopedIndent _s(line_appender_);
+#if 0
+      emit("// range known at runtime");
+      num_threads_ = 1; // TODO: grid-stride-loop
+      auto begin_expr = stmt->const_begin ? std::to_string(stmt->begin_value)
+        : fmt::format("_gtmp_i32_[{} >> 2]", stmt->begin_offset);
+      auto end_expr = stmt->const_end ? std::to_string(stmt->end_value)
+        : fmt::format("_gtmp_i32_[{} >> 2]", stmt->end_offset);
+      emit("int _tid = int(gl_GlobalInvocationID.x);");
+      emit("if (_tid >= {}) return;", num_threads_);
+      emit("int _beg = {}, _end = {};", begin_expr, end_expr);
+      emit("for (int _itv = _beg; _itv < _end; _itv++) {{");
+      {
+        ScopedIndent _s(line_appender_);
+        stmt->body->accept(this);
+      }
+      emit("}}\n");
+#else
+      emit("// range known at runtime");
+      auto begin_expr = stmt->const_begin ? std::to_string(stmt->begin_value)
+        : fmt::format("_gtmp_i32_[{} >> 2]", stmt->begin_offset);
+      auto end_expr = stmt->const_end ? std::to_string(stmt->end_value)
+        : fmt::format("_gtmp_i32_[{} >> 2]", stmt->end_offset);
+      emit("int _tid = int(gl_GlobalInvocationID.x);");
+      emit("int _beg = {}, _end = {};", begin_expr, end_expr);
+      emit("int _itv = _beg + _tid;");
+      emit("if (_itv >= _end) return;");
+      num_threads_ = -1;
+
+      std::vector<size_t> rinfo = {stmt->const_begin, stmt->const_end,
```

Comment:
nit: it's still better to do encapsulation here.. (https://testing.googleblog.com/2017/11/obsessed-with-primitives.html)  But I guess we can fix that later.

Reply:
Like the Metal's `KernelRangeAttrib`? Yes! Let's do that in another PR.